sympy/core/basic.py

```diff
@@ -1284,9 +1283,25 @@ def has_free(self, *patterns):
         True
         >>> (x + y + 1).has_free(y + 1)
         True
+        >>> (2*x*y + 3).has_free(x*y)
+        True
 
+        Performance is best when passing multiple patterns as a set.
         """
-        return self._has(iterfreeargs, *patterns)
+        if not patterns:
+            return False
+        if len(patterns) == 1:
+            if iterable(patterns[0]):
+                patterns = patterns[0]
```

Comment:
Checking iterable like this is problematic:
```python
In [7]: iterable(FiniteSet(1, 2, 3))
Out[7]: True

In [8]: iterable(Matrix([1]))
Out[8]: True
```
In fact I think that overloading `*args` vs ordinary args is usually a mistake to be avoided. We should just have a separate `has_free_set` method to do this.

Reply:
In case any Expr has an iterable for an arg:
```python
Piecewise((FiniteSet(1,2),x>1),(2,True)).has_free(FiniteSet(1,2))
```